clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -161,6 +166,39 @@ public synchronized int requestUpdate() {
         return this.version;
     }
 
+    /**
+     * Request an update for the current metadata info of a partition and ensure it is as recent as the given epoch,
+     * return the current version before the update
+     */
+    public synchronized int maybeRequestUpdate(TopicPartition topicPartition, int epoch) {
```

Comment:
I wonder if it would be useful to have this return a boolean so that we know whether or not an update was requested. Also, if a newer epoch is requested, should we somehow invalidate the current cached metadata for the partition? We wouldn't want to use the stale metadata for fetch requests, for example. This is the KIP-232 problem.